drivers/timer/nrf_rtc_timer.c

```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
 
 u32_t _timer_cycle_get_32(void)
 {
-	u32_t elapsed_cycles;
 	u32_t sys_clock_tick_count;
-	u32_t rtc_prev;
-	u32_t rtc_now;
+	u32_t rtc_count;
+	u32_t ticked_cycles;
+	u32_t elapsed_cycles;
 
-	rtc_now = RTC_COUNTER;
-	/* Discard value of RTC_COUNTER read at LFCLK transition */
-	do {
-		sys_clock_tick_count = _sys_clock_tick_count;
-		elapsed_cycles = (rtc_now - (sys_clock_tick_count *
-					     sys_clock_hw_cycles_per_tick)) &
-				 RTC_MASK;
-		rtc_prev = rtc_now;
-		rtc_now = RTC_COUNTER;
-	} while (rtc_now != rtc_prev);
```

Comment:
I don't think that such "protection" gives any value:

```
int get_time()
{
    int time;

/*
 * If RTC_REGISTER changes here, code below will get updated value.
 */
    do {
        time = RTC_REGISTER;
        /*
         *  If RTC_REGISTER changes here, the loop will actually delay return
         * from the function and returned value will include effect of the delay.
         */
     } while (time != RTC_REGISTER)

/*
 * If RTC_REGISTER changes here, value returned by this function will be off by 1 system clock tick.
 */

return time;
}
```

In both versions, the one with loop and the one without there is risk that a "stale" value will be returned. It is unavoidable, as even return from the function takes time and RTC counter could be updated during this time. I personally do not see any benefit from this loop (especially as in this case access to RTC is atomic and there is no risk that we get inconsistent value of the register).

Reply:
What you are not considering is the one clock tick that elapses, and the kernel cannot tolerate a stale clock tick. You can use your implementation on nRF52 boards and run a typical peripheral sample, it will stall eventually when the system clock ticks between the reading of the kernel variable and reading of the RTC counter.

Reply:
Are you mentioning here problem you pointed there: https://github.com/zephyrproject-rtos/zephyr/pull/9275#discussion_r207704157 ? If so, please look on my comment posted here: https://github.com/zephyrproject-rtos/zephyr/pull/9275#discussion_r207866959

Reply:
I understand your explanation. except for:
/*
 * If RTC_REGISTER changes here, value returned by this function will
 *  be off by 1 system clock tick.
 */

Here the returned value is only off by 1 cycle, not 1 system clock tick.

But as long as, tests cases pass as before, I am ok with the changes (because I am not very well informed on the kernel's use of this function).